Starting Worker for Realtime Adapters

gokulavasan opened the pull request:
Implementing Worker Adapters. 

Starting an Adapter which corresponds to a Worker ProgramType will start a run of that specific Worker program with the runtime arguments from the AdapterSpecification. It will also persist the information of adapter name -> RunId map in AdapterStore. This is required in order to stop the adapter which can be achieved only by knowing the RunId.

Also, a preliminary work on ProgramLifecycleService is part of this PR. This will abstract away the logic associated with starting a program and thus can be used in different places where Programs are started in CDAP - ProgramLifecycleHttpHandler, ScheduleTaskRunner, AdapterService.

Bamboo : http://builds.cask.co/browse/CDAP-DUT1375-9
JIRA : https://issues.cask.co/browse/CDAP-2050

albertshau commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/adapter/AdapterServiceTests.java:
```diff
@@ -74,18 +75,49 @@ public void testInvalidAdapter() throws Exception {
+
+    // Delete Adapter
+    adapterService.removeAdapter(NAMESPACE, adapter1);
+    adapterService.removeAdapter(NAMESPACE, adapter2);
```
verify they are deleted after removing

albertshau commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java:
```diff
@@ -0,0 +1,187 @@
+  private Program getProgram(Id.Program id, ProgramType programType) throws IOException {
+    Program program = store.loadProgram(id, programType);
+    if (program == null) {
+      throw new FileNotFoundException(String.format("Program not found: Id = %s; Type = %s", id, programType));
```
is this supposed to be a FileNotFound?

albertshau commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java:
```diff
@@ -373,12 +388,75 @@ private void stopWorkflowAdapter(Id.Namespace namespace, AdapterSpecification ad
+  private Id.Program getWorkerId(Id.Namespace namespace, AdapterSpecification adapterSpec) throws NotFoundException {
+    Id.Application appId = Id.Application.from(namespace, adapterSpec.getTemplate());
+    ApplicationSpecification appSpec = store.getApplication(appId);
+    if (appSpec == null || appSpec.getWorkers().size() != 1) {
```
these 2 errors are different and should be handled differently i think.  though how can the second case happen?  It seems like if it happens it's a bug in our code?

albertshau commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterStore.java:
```diff
@@ -0,0 +1,67 @@
+ * Store for Adapter related information.
+ */
+//TODO: Combine simple ConfigStore abstractions together.
+public class AdapterStore {
```
dont really like that state is split up into multiple stores.  It makes it harder to make sure things are consistent and correct.  I think it would be better to store this information where status is stored.

albertshau replied:
actually, how is current run information stored?  ideally we would be able to get run history for an adapter as well instead of just storing the current runid only for workers.

albertshau commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java:
```diff
@@ -373,12 +388,75 @@ private void stopWorkflowAdapter(Id.Namespace namespace, AdapterSpecification ad
+      store.setWorkerInstances(workerId, adapterSpec.getInstances());
+      ProgramRuntimeService.RuntimeInfo runtimeInfo = lifecycleService.start(workerId, ProgramType.WORKER,
+                                                                             sysArgs, userArgs, false);
+      adapterStore.setRunId(adapterId, runtimeInfo.getController().getRunId());
```
not introduced by you, but we can have races where requests to start the same program are made concurrently and then things break.  should probably have some sort of check and set logic on program status before we do anything.  Can you open a jira and add a todo somewhere?

albertshau commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java:
```diff
@@ -373,12 +388,75 @@ private void stopWorkflowAdapter(Id.Namespace namespace, AdapterSpecification ad
+      adapterStore.setRunId(adapterId, runtimeInfo.getController().getRunId());
+      final ProgramController controller = runtimeInfo.getController();
+      controller.addListener(new AbstractListener() {
+        @Override
```
i think we'll also want to override init and set the adapter status there instead of afterwards to reduce the chance of a race.

gokulavasan commented:
@albertshau Updated the PR. Addressed your comments. Please review. 
I have filed https://issues.cask.co/browse/CDAP-2112. Once that is resolved, I will add a test for RealtimeAdapterLifecycle to AdapterLifecycleTests.

albertshau commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java:
```diff
@@ -1416,88 +1414,21 @@ private AppFabricServiceStatus start(final Id.Program id, ProgramType type,
     } catch (Throwable throwable) {
       LOG.error(throwable.getMessage(), throwable);
-      if (throwable instanceof FileNotFoundException) {
+      if (throwable instanceof ProgramNotFoundException) {
```
should just have a separate catch for this exception instead of checking the type here. Also be sure not to log an error for this case.

albertshau commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java:
```diff
@@ -379,12 +395,88 @@ private void stopWorkflowAdapter(Id.Namespace namespace, AdapterSpecification ad
+        }
+      }, Threads.SAME_THREAD_EXECUTOR);
+    } catch (Throwable t) {
+      if (t instanceof FileNotFoundException) {
```
catch this separately. Shouldn't be FileNotFound right?

albertshau commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java:
```diff
@@ -379,12 +395,88 @@ private void stopWorkflowAdapter(Id.Namespace namespace, AdapterSpecification ad
+      if (t instanceof FileNotFoundException) {
+        throw new NotFoundException(workerId);
+      } else {
+        Throwables.propagate(t);
```
should try to avoid doing this whenever possible.  It changes checked exceptions to runtime exceptions means you lose the ability to do proper error handling.  why not include them as exceptions the method can throw?

albertshau commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java:
```diff
@@ -379,12 +395,88 @@ private void stopWorkflowAdapter(Id.Namespace namespace, AdapterSpecification ad
+    try {
+      lifecycleService.stopProgram(runId);
+    } catch (Throwable t) {
+      LOG.error("Error while trying to stop Adapter {} which has RunId of {} : ", adapterId, runId, t);
```
similar comment as above. Catching just to log an error and then propagate means callers can't handle errors differently.

albertshau commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskRunner.java:
```diff
@@ -136,26 +113,21 @@ public ScheduleTaskRunner(Store store, ProgramRuntimeService runtimeService, Pre
+    try {
+      runtimeInfo = lifecycleService.start(id, type, sysArgs, userArgs, false);
+    } catch (ProgramNotFoundException e) {
+      throw new TaskExecutionException(UserMessages.getMessage(UserErrors.PROGRAM_NOT_FOUND), e, false);
```
think you should just have the method throw ProgramNotFoundException instead of converting it to TaskExecutionException.  Seems like they are two very different types of errors.

gokulavasan replied:
The Javadoc mentions that TaskExecutionException is thrown if program is already running or program is not found.

albertshau commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java:
```diff
@@ -379,12 +395,88 @@ private void stopWorkflowAdapter(Id.Namespace namespace, AdapterSpecification ad
+                                               adapterSpec.getName());
+    RunRecord adapterRun = Iterables.getFirst(runRecords, null);
+    if (adapterRun == null) {
+      throw new NotFoundException(adapterId);
```
throwing this is misleading, stopAdapter documents that NotFound is thrown when the adapter could not be found, which means this will result in a 404 in the handler.  Should probably change the current NotFoundExceptions to AdapterNotFoundExceptions.

albertshau replied:
also, how can this happen?  If the run is not there should this just be a no-op instead of an error?

albertshau commented:
happy path logic looks good, most comments are around error scenarios

gokulavasan commented:
@albertshau Addressed comments. Please review.

albertshau commented:
:+1:
